paper-server/patches/sources/net/minecraft/server/network/ServerGamePacketListenerImpl.java.patch

```diff
@@ -2620,6 +2622,11 @@
  
      @Override
      public void handleClientTickEnd(ServerboundClientTickEndPacket packet) {
+         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel());
++        this.tickEndEvent.callEvent(); // Paper - add client tick end event
```

Comment:
Please check that the handler list is > 0 before executing this. 
Just cause of how often this will be running.

Reply:
I thought about that, but since we're re-using the event instance checking the handler list size wouldn't really avoid anything, one thread check at most